flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/RocksDBRocksIteratorWrapperTest.java

```diff
@@ -0,0 +1,147 @@
+
+import static org.mockito.Mockito.mock;
+
+/**
+ * Tests for the RocksIteratorWrapper.
+ */
+public class RocksDBRocksIteratorWrapperTest {
+
+	@Rule
+	public final TemporaryFolder tmp = new TemporaryFolder();
+
+	@Test
+	public void testIterator() throws Exception{
+
+		// test for keyGroupPrefixBytes == 1 && ambiguousKeyPossible == false
+		testIteratorHelper(IntSerializer.INSTANCE, StringSerializer.INSTANCE, 128, i -> i);
+
+		// test for keyGroupPrefixBytes == 1 && ambiguousKeyPossible == true
+		testIteratorHelper(StringSerializer.INSTANCE, StringSerializer.INSTANCE, 128, i -> String.valueOf(i));
+
+		// test for keyGroupPrefixBytes == 2 && ambiguousKeyPossible == false
+		testIteratorHelper(IntSerializer.INSTANCE, StringSerializer.INSTANCE, 256, i -> i);
+
+		// test for keyGroupPrefixBytes == 2 && ambiguousKeyPossible == true
+		testIteratorHelper(StringSerializer.INSTANCE, StringSerializer.INSTANCE, 256, i -> String.valueOf(i));
+	}
+
+	<K> void testIteratorHelper(
+		TypeSerializer<K> keySerializer,
+		TypeSerializer namespaceSerializer,
+		int maxKeyGroupNumber,
+		Function<Integer, K> getKeyFunc) throws Exception {
+
+		String testStateName = "aha";
+		String namespace = "ns";
+
+		String dbPath = tmp.newFolder().getAbsolutePath();
+		String checkpointPath = tmp.newFolder().toURI().toString();
+		RocksDBStateBackend backend = new RocksDBStateBackend(new FsStateBackend(checkpointPath), true);
+		backend.setDbStoragePath(dbPath);
+
+		Environment env = new DummyEnvironment("TestTask", 1, 0);
+
+		RocksDBKeyedStateBackend<K> keyedStateBackend = (RocksDBKeyedStateBackend<K>) backend.createKeyedStateBackend(
+			env,
+			new JobID(),
+			"Test",
+			keySerializer,
+			maxKeyGroupNumber,
+			new KeyGroupRange(0, maxKeyGroupNumber - 1),
+			mock(TaskKvStateRegistry.class));
+
+		keyedStateBackend.restore(null);
+
+		ValueState<String> testState = keyedStateBackend.getPartitionedState(
+			namespace,
+			namespaceSerializer,
+			new ValueStateDescriptor<String>(testStateName, String.class));
+
+		// insert record
+		for (int i = 0; i < 1000; ++i) {
+			keyedStateBackend.setCurrentKey(getKeyFunc.apply(i));
+			testState.update(String.valueOf(i));
+		}
+
+		ColumnFamilyHandle handle = keyedStateBackend.getColumnFamilyHandle(testStateName);
+		RocksIterator iterator = keyedStateBackend.db.newIterator(handle);
+		iterator.seekToFirst();
+
+		ByteArrayOutputStreamWithPos outputStream = new ByteArrayOutputStreamWithPos(8);
+		boolean ambiguousKeyPossible = AbstractRocksDBState.AbstractRocksDBUtils.isAmbiguousKeyPossible(keySerializer, namespaceSerializer);
+		AbstractRocksDBState.AbstractRocksDBUtils.writeNameSpace(
+			namespace,
+			namespaceSerializer,
+			outputStream,
+			new DataOutputViewStreamWrapper(outputStream),
+			ambiguousKeyPossible);
+
+		byte[] nameSpaceBytes = outputStream.toByteArray();
+
+		RocksDBKeyedStateBackend.RocksIteratorWrapper<K> iteratorWrapper = new RocksDBKeyedStateBackend.RocksIteratorWrapper(
```

Comment:
Same here, this is never closed. A test should also check that calling the wrapper's close closes the native iterator.

Reply:
Nice catch! Addressing